File: app/src/main/java/com/zxy/demo/CustomView.java
package com.zxy.demo;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.Scroller;

import com.zxy.utility.LogUtil;

/**
 * Created by zxy on 2018/9/14 09:55.
 */

public class CustomView extends View
{
    Paint mPaint;
    Scroller mScroller;

    {
        mScroller = new Scroller(getContext());
        mScroller.startScroll(0, 0, 300, 0, 9000);
        invalidate();
    }

    public CustomView(Context context)
    {
        super(context);
    }

    public CustomView(Context context, @Nullable AttributeSet attrs)
    {
        super(context, attrs);
    }

    public CustomView(Context context, @Nullable AttributeSet attrs, int defStyleAttr)
    {
        super(context, attrs, defStyleAttr);
    }


    @Override
    protected void onDraw(Canvas canvas)
    {
        super.onDraw(canvas);
        mPaint = new Paint();
        mPaint.setColor(Color.RED);
        mPaint.setAntiAlias(true);
        canvas.drawColor(Color.GRAY);
        DisplayMetrics dm = new DisplayMetrics();
        ((WindowManager)getContext().getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay().getMetrics(dm);
        canvas.drawCircle(dm.widthPixels/2, dm.heightPixels/2, 100.f, mPaint);

    }

    private float mStartX;
    private float mStartY;

    private float mEndX;
    private float mEndY;

    @Override
    public void computeScroll()
    {
        if (mScroller.computeScrollOffset())
        {
            LogUtil.e("scrollOffset:StartX" + mScroller.getStartX() + " StartY:" + mScroller.getStartY()
                    + "\n CurVelocity:" + mScroller.getCurrVelocity()
                    + "\n CurX:" + mScroller.getCurrX() + " CurY:" + mScroller.getCurrY()
                    + "\n FinalX:" + mScroller.getFinalX() + " FinalY:" + mScroller.getFinalY()
                    + "\n dura:" + mScroller.getDuration());
            scrollTo(-mScroller.getCurrX(),mScroller.getCurrY());
        }
        super.computeScroll();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event)
    {

        mStartX = event.getX();
        mStartY = event.getY();
        switch (event.getAction())
        {
            case MotionEvent.ACTION_DOWN:

                break;
            case MotionEvent.ACTION_MOVE:
                break;
            case MotionEvent.ACTION_UP:

                break;
        }
        return true;
    }
}

File: app/src/main/java/com/zxy/demo/CustomView.kt
package com.zxy.demo

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

/**
 *@Description
 *@author          Administrator
 *@create          2021-03-29 17:33
 */
class CustomView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {

    val paint = Paint()

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        paint.setStrokeWidth(20f)
//        paint.setStrokeCap(Paint.Cap.SQUARE)
//        canvas?.drawPoint(50f, 50f, paint)
//
//        val points = floatArrayOf(0f, 0f, 50f, 50f, 50f, 100f, 100f, 50f, 100f, 100f, 150f, 50f, 150f, 100f)
//// 绘制四个点：(50, 50) (50, 100) (100, 50) (100, 100)
//// 绘制四个点：(50, 50) (50, 100) (100, 50) (100, 100)
//        canvas!!.drawPoints(points, 1 /* 跳过两个数，即前两个 0 */,
//                8 /* 一共绘制 8 个数（4 个点）*/, paint)
//        paint.setStyle(Paint.Style.FILL);
//        canvas.drawOval(50f, 50f, 350f, 200f, paint);

//        val points = floatArrayOf(20f, 20f, 120f, 20f, 70f, 20f, 70f, 120f, 20f, 120f, 120f, 120f, 150f, 20f, 250f, 20f, 150f, 20f, 150f, 120f, 250f, 20f, 250f, 120f, 150f, 120f, 250f, 120f)
//        canvas?.drawLines(points, paint)
//        canvas?.drawRoundRect(100f, 100f, 500f, 300f, 50f, 50f, paint);

        paint.setStyle(Paint.Style.FILL); // 填充模式
        canvas?.drawArc(200f, 100f, 800f, 500f, -110f, 100f, true, paint);
        canvas?.drawArc(200f, 100f, 800f, 500f, 180f, 60f, false, paint);
    }
}